Interpreter Fixes

fmatthew5876 opened the pull request:
Fixes: #1565

fmatthew5876 commented:
Fixes #1559 and implements 1 commit from #1401

fmatthew5876 commented:
All of the calls are immediate except random encounter battles and menu triggered by `ESC`.

The non-immediate cases for menu and battle are not yet perfectly frame accurate but with this branch they are pretty close and exhibit the main features. The main one being esc menu waits for you to finish you step before triggering.

Ghabry commented on src/game_map.cpp:
```diff
@@ -128,6 +128,7 @@ void Game_Map::Dispose(bool is_load_savegame) {
 
 	map.reset();
 	animation.reset();
+	interpreter.reset(new Game_Interpreter_Map(0, true));
```
Please edit the issue number this fixes in the commit

Ghabry replied:
also you can't simply reset the global Interpreter on map change, events from the previous map continue.

fmatthew5876 replied:
Fixed

Ghabry commented:
Same aplies to other commits that has numbers.

Interpreter is very tricky, so we need good documentation here in case some lines cause regressions later.

Ghabry commented on src/game_interpreter_map.h:
```diff
@@ -38,6 +38,8 @@ class Game_Interpreter_Map : public Game_Interpreter
 public:
 	Game_Interpreter_Map(int _depth = 0, bool _main_flag = false);
 
+	~Game_Interpreter_Map() {}
```
any reason for this?
I bet some obscure corner case specified in the standard almost nobody knows for "C++ will leak this because it is inherited"?

Ghabry commented on src/scene_map.cpp:
```diff
@@ -260,14 +266,12 @@ void Scene_Map::CallBattle() {
 }
 
 void Scene_Map::CallShop() {
-	Game_Temp::shop_calling = false;
```
Any reason the Game_Map::GetInterpreter().ResetImmediateCalling is in Game_Interpreter and not in the Call-functions as before?

Ghabry commented on src/scene_map.cpp:
```diff
@@ -190,7 +190,7 @@ void Scene_Map::Update() {
 
 	if (!Main_Data::game_player->IsMoving() || Game_Map::GetInterpreter().IsImmediateCall()) {
-		if (Game_Temp::menu_calling || Game_Temp::menu_calling_immediate) {
+		if (Main_Data::game_data.party_location.menu_calling || Game_Temp::menu_calling_immediate) {
```
Maybe Main_Data::game_data.party_location could be exposed via functions in Game_Party?

fmatthew5876 commented:
I've refactored this:

* The whole `Game_Map::Dispose(bool)` thing is ugly. I changed it to call `Clear()` inside `Game_Interpreter_Map::SetupFromSave()`.
* Moved the `Game_Temp` calling flags (except battle) into `Game_Interpreter`. Instead of trying to rebase them and messing up, I've smashed the sormat, menu_calling, and game_temp features into 1 commit. Apologies if its a bit large.

fmatthew5876 commented on src/game_interpreter.cpp:
```diff
@@ -197,11 +198,12 @@ void Game_Interpreter::Update() {
+			IsMenuCalling() ||
+			IsSaveCalling() ||
+			IsLoadCalling() ||
+			Main_Data::game_data.party_location.menu_calling ||
```
Not sure if we need the ESC menu calling flag here.

@Ghrabry what do you think about this?

fmatthew5876 replied:
Fixed

Ghabry commented:
Attack on Game_Temp has begun, 5 variables removed 👍

Ghabry commented:
Jenkins: test this please
